### PR TITLE
CUPTI support

### PR DESCRIPTION
**Pull request by senior-zero**
Initial CUPTI support enables collection of the following metrics:
- L1 hit rates
- L2 hit rates
- DRAM throughput
- Store efficiency
- Load efficiency

This addresses [the following issue](https://github.com/NVIDIA/nvbench/issues/5). I haven't enabled CUPTI by default because there'll be a crash on Windows systems if the CUPTI lib64 directory isn't in PATH. To enable CUPTI support, it's sufficient to add `-DNVBench_ENABLE_CUPTI=ON`. After that, it's possible to specify that any of the following metrics are required:

```cuda
state.collect_dram_throughput();
state.collect_l1_hit_rates();
state.collect_l2_hit_rates();
state.collect_loads_efficiency();
state.collect_stores_efficiency();
```

I've added an example `examples/auto_throughput.cu` which produces the following output on both Linux and Windows:

| T | Stride | Elements | HBWPeak | LoadEff | StoreEff | L1HitRate | L2HitRate | Samples |  CPU Time  | Noise |  GPU Time  | Noise |  Elem/s  | Batch GPU  | Batch |
|---|--------|----------|---------|---------|----------|-----------|-----------|---------|------------|-------|------------|-------|----------|------------|-------|
| 1 |      1 | 33554432 |  91.35% | 100.00% |  100.00% |     0.00% |    50.02% |     11x | 359.400 us | 0.80% | 317.719 us | 0.15% | 105.610G | 318.293 us |   12x |
| 1 |      4 | 33554432 |  94.87% |  25.00% |  100.00% |     0.00% |    20.03% |     11x | 814.845 us | 0.34% | 773.027 us | 0.04% |  43.407G | 831.232 us |   12x |
| 2 |      1 | 33554432 |  85.71% |  50.00% |  100.00% |    30.17% |    51.88% |     11x | 737.191 us | 2.48% | 694.644 us | 2.72% |  48.304G | 747.861 us |   12x |
| 2 |      4 | 33554432 |  75.09% |  12.50% |  100.00% |    35.65% |    28.64% |     11x |   2.060 ms | 1.36% |   2.018 ms | 1.38% |  16.630G |   2.170 ms |   12x |

### CODE REVIEW

File: nvbench/CMakeLists.txt
Comment (allisonvacanti) [inline nvbench/CMakeLists.txt:74]: @robertmaynard Do you know anything about this?

@senior-zero mentioned that he had to modify `PATH` like this when linking to CUPTI on windows. Is there a way to fix this without bugging the user?

Comment (robertmaynard) [reply nvbench/CMakeLists.txt:74]: Link to the `CUDA::cuda_driver` and `CUDA::cupti` targets instead of having the explicit paths in target_include_directories, and target_link_directories plus the usage of raw libraries ( cuda, cupti ) in target_link_libraries

File: nvbench/detail/measure_cupti.cu
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cu:228]: You may find the `NVBENCH_THROW` macro in `nvbench/detail/throw.cuh` helpful here :)

File: nvbench/detail/measure_cupti.cu
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cu:244]: anon namespace, similar naming suggestion as the other metric helpers (`gen_summaries_impl`)

File: nvbench/detail/measure_cupti.cu
Comment (allisonvacanti) [reply nvbench/detail/measure_cupti.cu:36]: Use a scoped enum for improved type-safety: `enum class metric_ids`.

We should also decide whether we want enum scope names to be plural or singular, e.g. `metric_ids` vs. `metric_id`. I think I prefer singular, but I haven't been consistent about this...

File: nvbench/detail/measure_cupti.cu
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cu:279]: Consider the function scope try/catch syntax when the entire function is in a try block:

```
void function()
try {
  ...
}
catch (...)
{
  ...
}
```

File: nvbench/detail/measure_cupti.cuh
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cuh:26]: alphabetize

File: nvbench/detail/measure_cupti.cuh
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cuh:79]: Why optional?

File: nvbench/device_info.cuh
Comment (allisonvacanti) [inline nvbench/device_info.cuh:187]: Some API capitalizes `CUPTI`, some just uses lowercase `cupti`. Let's just stick with lowercase throughout the API, since other abbreviations are handled that way.

File: nvbench/device_info.cuh
Comment (allisonvacanti) [inline nvbench/device_info.cuh:69]: `NVBENCH_HAS_CUPTI`

File: nvbench/device_info.cuh
Comment (allisonvacanti) [inline nvbench/device_info.cuh:189]: `NVBENCH_HAS_CUPTI`

File: nvbench/device_info.cuh
Comment (allisonvacanti) [inline nvbench/device_info.cuh:196]: `NVBENCH_HAS_CUPTI`

File: nvbench/main.cuh
Comment (allisonvacanti) [inline nvbench/main.cuh:48]: `NVBENCH_HAS_CUPTI`

File: nvbench/detail/state_exec.cuh
Comment (allisonvacanti) [inline nvbench/detail/state_exec.cuh:80]: Use explicit `this->` for member function calls.

File: nvbench/detail/state_exec.cuh
Comment (allisonvacanti) [inline nvbench/detail/state_exec.cuh:99]: `this->is_cupti_required()`

File: nvbench/CMakeLists.txt
Comment (allisonvacanti) [inline nvbench/CMakeLists.txt:28]: Rather than unconditionally compiling these and slicing bits of them out with `NVBENCH_HAS_CUPTI`, let's do this:

```
if (NVBench_ENABLE_CUPTI)
  list(APPEND srcs [cupti files here])
endif()
```

and then remove the `NVBENCH_HAS_CUPTI` checks from these files. This clean separation will reduce the complexity of maintaining the optional dependency, since the only significant code divergence will be in `state_exec.cuh`.

We'll need to move error logging for the "cupti is not enabled but requested" case in `state_exec.cuh` with this approach.

Comment (allisonvacanti) [reply nvbench/CMakeLists.txt:28]: Don't list the header files here -- it's harmless, but currently we're only listing translation units in the `srcs` list.

File: nvbench/CMakeLists.txt
Comment (allisonvacanti) [inline nvbench/CMakeLists.txt:31]: Use the `.cxx` extension for consistency.

File: nvbench/CMakeLists.txt
Comment (allisonvacanti) [reply nvbench/CMakeLists.txt:78]: This may also fix the windows PATH issue above.

File: nvbench/cupti_profiler.cpp
Comment (allisonvacanti) [inline nvbench/cupti_profiler.cpp:81]: Why throw away exceptions here?

File: nvbench/detail/measure_cupti.cu
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cu:201]: Should this be logged?

File: nvbench/detail/measure_cupti.cu
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cu:283]: Do we really want to throw away exceptions here?

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
This is looking great! Really nice addition -- most of my comments are minor and repetitive ;)

File: nvbench/CMakeLists.txt
Comment (allisonvacanti) [reply nvbench/CMakeLists.txt:78]: Looks like Rob confirmed this while I was finishing the review 😁

File: nvbench/CMakeLists.txt
Comment (allisonvacanti) [reply nvbench/CMakeLists.txt:74]: #46 adds NVML support and fixes a lot of the CMake issues in this PR:

- Fixes/WARs the dll PATH issues without requiring users to modify PATH
- Adds an auto-generated `nvbench/config.cuh` header instead of relying on compiler definitions.

It should be fairly easy to extend the approaches used in that PR here. I'd suggest rebasing this on #46 so we can reuse the cmake machinery.

File: nvbench/cupti_profiler.cpp
Comment (senior-zero) [reply nvbench/cupti_profiler.cpp:128]: Judging on CUPTI samples, it's using CUDA ids.

File: nvbench/cupti_profiler.cpp
Comment (senior-zero) [reply nvbench/cupti_profiler.cpp:602]: I'm not sure I understand. Which case are you referring to? https://godbolt.org/z/rd8zfcMKW 
Neither bool nor std::vector has `[[nodiscard]]` in their declarations. Having `[[nodiscard]]` in the `cupti_profiler` wouldn't affect it's methods.

Comment (allisonvacanti) [reply nvbench/cupti_profiler.cpp]: The `[[nodiscard]]` attribute is already used inside the definition of `cupti_profiler` on the declaration of the `cupti_profiler::initialize` method:

```
struct cupti_profiler
{
  // ...
  [[nodiscard]] bool initialized() const;
  // ...
};
```

and it's not necessary to repeat it here on the method definition. Ex: https://godbolt.org/z/KMT8hhW6v

I prefer to just have these attributes inside the class definition to avoid having to keep both decl/def sync'd, but it's admittedly an extremely minor nitpick 🙂

File: nvbench/detail/measure_cupti.cu
Comment (allisonvacanti) [reply nvbench/detail/measure_cupti.cu]: For reference, here's how to print a warning to the log: https://github.com/NVIDIA/nvbench/blob/main/nvbench/detail/measure_cold.cu#L158-L196

File: nvbench/detail/state_exec.cuh
Comment (allisonvacanti) [inline nvbench/detail/state_exec.cuh:31]: When I run `nvbench.example.auto_throughput` with CUPTI disabled, I get

```
bin/nvbench.example.auto_throughput: symbol lookup error: bin/nvbench.example.auto_throughput: undefined symbol: nvbench::detail::measure_cupti_base::generate_summaries()
```

Wrapping this header in `#ifdef NVBENCH_HAS_CUPTI / #endif` fixes it.

File: nvbench/cupti_profiler.cuh
Comment (allisonvacanti) [inline nvbench/cupti_profiler.cuh:42]: This example is very helpful!

Small language nitpick about this method name, `prepare_user_loop` sounds more natural.

File: nvbench/detail/measure_cupti.cuh
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cuh:92]: This is redundant with line 23.

I think we can probably remove all of these checks and just ifdef out the `#include <nvbench/detail/measure_cupti.cuh>` in `state_exec.cuh`.

File: nvbench/detail/measure_cupti.cuh
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cuh:105]: Same note as above.

File: nvbench/detail/measure_cupti.cuh
Comment (allisonvacanti) [inline nvbench/detail/measure_cupti.cuh:138]: Same as above.

File: nvbench/device_info.cuh
Comment (allisonvacanti) [inline nvbench/device_info.cuh:226]: Move this function into `device_info.cu` (to avoid leaking fmt) and use the NVBENCH_THROW macro to embed file/line information.

File: nvbench/main.cuh
Comment (allisonvacanti) [inline nvbench/main.cuh:50]: Prefix this with `NVBENCH_` since it needs to leak into user code.

**Review by allisonvacanti — APPROVED** [review]
A couple minor issues in the comments, but otherwise I'm really happy with this -- it's a very useful addition and I appreciate your hard work here!

Approving, once nvml is merged and these last comments are address this is ready to go in. 👍

**Pull request merged**
